src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1801,7 +1801,21 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             SyntaxNode syntax,
             bool isCall)
         {
-            if (Join(ref localFunctionState.StartingState, ref State) &&
+            // When we are in a conditional state after analyzing a call to a local function,
+            // e.g. when the local function uses a conditional flow analysis attribute,
+            // we just analyze the local function usage using a temporary joined state.
+            LocalState state;
+            if (IsConditionalState)
+            {
+                state = StateWhenTrue.Clone();
+                state.Join(in StateWhenFalse);
+            }
+            else
+            {
+                state = State;
+            }
+
+            if (Join(ref localFunctionState.StartingState, ref state) &&
```

Comment:
>if (Join(ref localFunctionState.StartingState, ref state) && [](start = 12, length = 60)

Perhaps I am just confused by the terminology used here, but it feels like the ```state``` is a state after the call,  whereas StartingState implies it is a state on entry into the function, i.e. before the call. Why would we want to "Join" them?

Reply:
/cc @agocke in case you want to weigh in here.

It seems like the state we want to join into the local function's StartingState is one where all arguments have been visited, but postconditions from the call have not yet been applied.

The current implementation can cause diagnostics which seem undesirable:
```cs
void M()
{
    var s = ""hello"";

    local1(out s);

    bool local1([MaybeNullWhen(false)] out string s1)
    {
        s.ToString(); // currently warns, but probably shouldn't
        s1 = ""world"";
        return true;
    }
}
```

Reply:
> It seems like the state we want to join into the local function's StartingState is one where all arguments have been visited, but postconditions from the call have not yet been applied.

Good point. Maybe we should call `VisitLocalFunctionUse` at the start of `VisitCall` rather than as we currently do (at the end of `VisitCall`.

That said, I'm not sure if that's Aleksey's question. 
The way I understand it: the reason why we `Join` is that we want to get the worst case state at the start of all local function invocations, then we'll analyze the local function with that state.

Reply:
It also seems like a similar bug is permitting us to use variables which have not been definitely assigned. I will try to fix both issues in this PR.

```cs
void M()
{
    string s0;
    local1(out s0);

    void local1(out string s1)
    {
        s0.ToString(); // should produce an error, but doesn't
        s1 = ""hello"";
    }
}
```

Reply:
It seems like a little refactoring is needed to fix the definite assignment issue. Will file a separate issue and fix in a different PR.

Reply:
Filed #41631